Refactor order_processor to be less dependent to db transactions

ashkan18 opened the pull request:
# Problem
As we started building on top of existing code around submitting an `Order`, things got more complicated. Looking deeper at our process and after few brainstorming, it seems the part that complicates things the most is us having most of business logic during submit in a `order.submit!` or `order.approve!` block. Anything in that block is using `order.with_lock` and is part of one database transaction.
While having everything during wrapped in one db transaction helps our data integrity, because of external dependencies like inventory deduction in Gravity and charge processing in Stripe, we can easily get into a rollback scenario. Rollback case also gets pretty complicated because in some cases we want to store transactions in some cases we don't.

# Solution
Instead of wrapping everything in a transaction block, do things in more sequential way and provide a proper rollback method in case of rollback in different stages.

In this new approach we rely more on `OrderProcessor` which used to be this weird little service only dealing with inventory deduction and charge processing. Now `OrderProcessor`:
- provides separate methods for charge related methods and deduct and undeducting inventory
- provides new methods for setting totals on the order
- provides a `rollback!` method that can decide what needs to be rolled back at current stage.

The big difference in the flow is now we optimistically set the order in new state at the beginning of the process and if anything goes wrong we rollback the state to previous one. We used to not be able to go from `submitted` state to `pending` but with new approach if order processing fails after our optimistic state update, we'll be going back from `submitted` to `pending` and in case of approve, from `approved` to `submitted`. This should be ok since we don't post events yet for any of these.


# How confident are we? %87.3
As long as this PR doesn't change any request level spec, i feel %87.3 confident. The rest of confidence happens when we see things in real 🌏 with our own 👁.

https://twitter.com/i/status/1139548144390352897

## Photo of the rest of %12.7
![](https://media.giphy.com/media/1rf4hhXCoTQNa/giphy.gif)

ashkan18 commented:
Still need to update some service layer tests but this should be reviewable at this point.

joeyAghion commented on app/services/offer_service.rb:
```diff
@@ -66,33 +66,29 @@ def self.accept_offer(offer, user_id)
+      order_processor.rollback!
+      order_processor.set_payment!
+      raise Errors::PaymentRequiresActionError.new(order_processor.action_data)
+    else
```
I'm not sure you need this `else`. Since both the above conditions break the control flow by raising exceptions, you could just move the following 2 lines to outside of the `if`.

joeyAghion commented on app/services/order_service.rb:
```diff
@@ -48,40 +48,29 @@ def self.submit!(order, user_id)
-
-      order.update!(external_charge_id: order_processor.transaction.external_id)
+    order_processor.transition(:submit!)
+    order_processor.deduct_inventory!
```
This too looks like it would raise an error, but doesn't.

Minor refactoring idea: there are several places where you call a method and then call another to determine whether the first was successful. Instead, the first method could return `true` or `false` and the logic could be combined into lines like:

    unless order_processor.deduct_inventory do
      # rollback, etc.
    end

Or, if the logic is super-simple, even:

    order_processor.deduct_inventory || rollback...

joeyAghion commented on app/services/order_service.rb:
```diff
@@ -48,40 +48,29 @@ def self.submit!(order, user_id)
       Exchange.dogstatsd.increment 'submit.requires_action'
-      raise Errors::PaymentRequiresActionError, order_processor.action_data
+      raise Errors::PaymentRequiresActionError.new(order_processor.action_data)
+    else
```
Similar to above, could you get rid of this `else`?

joeyAghion commented on lib/order_processor.rb:
```diff
@@ -10,42 +10,63 @@ class OrderProcessor
+  end
+
+  def rollback!
+    undeduct_inventory! unless @deducted_inventory.empty?
```
Slight preference for the positive version: `if @deducted_inventory.any?`

joeyAghion commented on lib/order_processor.rb:
```diff
@@ -76,19 +97,36 @@ def valid?
+    PostTransactionNotificationJob.perform_later(transaction.id, @user_id)
+  end
+
+  def set_payment!
```
This is an unexpected name for a method that stores an `external_charge_id`. If it remains I'd consider renaming.

ashkan18 replied:
renamed it to `set_external_payment`

joeyAghion commented on lib/order_processor.rb:
```diff
@@ -76,19 +97,36 @@ def valid?
+    @order.update!(external_charge_id: transaction.external_id)
+  end
+
+  def set_follow_ups
```
The work below is only partially about follow-ups, so I might rename this to `on_submit` or `on_success`.

joeyAghion commented on lib/payment_service.rb:
```diff
@@ -158,6 +158,9 @@ def self.generate_transaction_from_exception(exc, type, external_id: nil, extern
     pi = body[:error][:payment_intent]
+    transaction_type = unless capture.nil?
+      capture ? Transaction::CAPTURE : Transaction::HOLD
+    end
```
A bit confusing! I don't have a great suggestion though.

ashkan18 replied:
oy.... this is merge conflict actually when i got latest master, the method already gets `transaction_type`, no need for this mess 🤦‍♂

starsirius commented on lib/order_processor.rb:
```diff
@@ -10,42 +10,63 @@ class OrderProcessor
+  end
+
+  def set_totals!
+    order.line_items.each { |li| li.update!(commission_fee_cents: li.current_commission_fee_cents) } if order.mode === Order::BUY
```
Is the use of `===` here (and below) intentional? I rarely use it in Ruby and, in terms of strictness, it seems to be the opposite of Javascript. _I think_ it's the the same when comparing two [String instances](https://ruby-doc.org/core-2.4.0/String.html#method-i-3D-3D-3D), but maybe using `==` is more explicit.

ashkan18 replied:
no! this must be the cause of using those weird fonts on my vscode! https://github.com/tonsky/FiraCode

starsirius commented on app/services/offer_service.rb:
```diff
@@ -66,33 +66,29 @@ def self.accept_offer(offer, user_id)
+    order_processor.transition(:approve!)
+    order_processor.deduct_inventory!
+    if order_processor.failed_inventory?
+      order_processor.rollback!
```
Maybe `transition` and `rollback` can pair better by renaming them to something like `advance_state(action)` and `revert_state`.

ashkan18 replied:
yes! love these new names!

ashkan18 replied:
@joeyAghion  i think i'd just call this `order_processor` method `revert` since it may not always revert the state

starsirius commented on lib/order_processor.rb:
```diff
@@ -10,42 +10,63 @@ class OrderProcessor
+  def rollback!
+    undeduct_inventory! unless @deducted_inventory.empty?
+    reset_totals! if @totals_set
+    order.rollback! if @state_changed
```
Do we want to unset `@state_changed`?

ashkan18 replied:
oh good point!

starsirius commented on lib/order_processor.rb:
```diff
@@ -10,42 +10,63 @@ class OrderProcessor
+
+  def reset_totals!
+    order.line_items.each { |li|  li.update!(commission_fee_cents: nil) } if order.mode === Order::BUY
+    order.update!(transaction_fee_cents: nil, commission_rate: nil, commission_fee_cents: nil, seller_total_cents: nil)
```
Similarly, maybe we want to unset `@totals_set`.

ashkan18 commented on app/services/offer_service.rb:
```diff
@@ -66,33 +66,29 @@ def self.accept_offer(offer, user_id)
-    ReminderFollowUpJob.set(wait_until: order.state_expiration_reminder_time).perform_later(order.id, order.state)
-    Exchange.dogstatsd.increment 'order.approved'
+    order_processor.set_totals!
+    order_processor.charge!
```
yes, we used `!` in Exchange also for methods that update the model, kind of like `merge` vs. `merge!`, that said, charge doesn't do that anymore though, so im going to remove `!`

ashkan18 commented on app/services/offer_service.rb:
```diff
@@ -66,33 +66,29 @@ def self.accept_offer(offer, user_id)
+      order_processor.set_payment!
+      raise Errors::PaymentRequiresActionError.new(order_processor.action_data)
+    else
+      order_processor.set_payment!
```
interesting, in that case, store transaction needs to know that if charge was failed don't update `external_charge_id` on the `order`, maybe name `set_payment` is wrong, maybe it should be `set_related_payment` or something?

ashkan18 commented:
@joeyAghion @starsirius addressed most of the comments, still adding more tests to `order_processor` but please take a look when had  a chance, would love to have this on staging sooner to unblock offer SCA works.

joeyAghion commented on app/models/order.rb:
```diff
@@ -66,7 +66,7 @@ class Order < ApplicationRecord
   ].freeze
 
-  ACTIONS = %i[abandon submit approve reject fulfill seller_lapse buyer_lapse refund].freeze
+  ACTIONS = %i[abandon rollback submit approve reject fulfill seller_lapse buyer_lapse refund].freeze
```
I'd still prefer a non-transaction-sounding name for this action.

joeyAghion commented on app/services/offer_service.rb:
```diff
@@ -66,33 +66,29 @@ def self.accept_offer(offer, user_id)
-      raise ActiveRecord::Rollback if order_processor.failed_payment?
-
-      order.update!(external_charge_id: order_processor.transaction.external_id)
+    order_processor.advance_state(:approve!)
```
The `:approve!` argument is a little strange. I'd expect a state or action name instead.

ashkan18 replied:
`approve` is the action name, do you mean removing `!` or make it string or both?

joeyAghion replied:
It's minor so feel free to do whatever's convenient here.

joeyAghion commented on app/services/offer_service.rb:
```diff
@@ -66,33 +66,29 @@ def self.accept_offer(offer, user_id)
-    Exchange.dogstatsd.increment 'order.approved'
+    order_processor.set_external_payment!
+    order_processor.on_success
+    order
```
Actually this is pretty clear now!

joeyAghion commented on app/services/order_service.rb:
```diff
@@ -48,40 +48,27 @@ def self.submit!(order, user_id)
-    OrderFollowUpJob.set(wait_until: order.state_expires_at).perform_later(order.id, order.state)
-    ReminderFollowUpJob.set(wait_until: order.state_expiration_reminder_time).perform_later(order.id, order.state)
-    Exchange.dogstatsd.increment 'order.submitted'
+    order_processor.set_external_payment!
```
I still wonder if the `set_external_payment` work could just be combined into `store_transaction`. `OrderProcessor` isn't a bad place for some logic about other columns to update based on the results of a transaction.

ashkan18 replied:
agreed, this may make next refactoring easier where i think we should switch to point to our own transaction instead of external id

joeyAghion commented on lib/order_processor.rb:
```diff
@@ -1,54 +1,62 @@
+    reset_totals! if @totals_set
+    return unless @state_changed
+
+    order.rollback!
```
I liked the suggestion from earlier of calling this `revert`.

ashkan18 replied:
yes, going to do this now.

ashkan18 commented:
alright! i actually addressed all comments now

ashkan18 commented:
alright... merging....
